rfcs/0009-defaults-and-overrides.md

```diff
@@ -0,0 +1,945 @@
+    block-semantics -- "Defaults" --> merge-default-block-into-p[[Merge default block<br>of policy rules <i>B</i><br>into policy <i>p</i>]] --> merge-block-of-rules-into-p-finish
+    block-semantics -- "Overrides" --> merge-override-block-into-p[[Merge override block<br>of policy rules <i>B</i><br>into policy <i>p</i>]] --> merge-block-of-rules-into-p-finish
+```
+
+### Merge a `defaults` block of policy rules into a policy
+
+```mermaid
+%%{ init: { "theme": "neutral" } }%%
+flowchart LR
+    merge-default-block-into-p-start([Merge default block<br>of policy rules <i>B</i><br>into policy <i>p</i>]) -->
+    unset-unwanted-policy-rules[Remove from <i>B</i><br>all policy rules<br>listed in <i>p.unset</i>] -->
+    p-empty{<i>p.empty?</i>}
+    p-empty -- "Yes" --> full-replace-p-with-defaut-block[<i>p.rules ← B</i>] --> merge-default-block-into-p-finish(((Return <i>p</i>)))
+    p-empty -- "No" --> default-block-strategy{<i>B.strategy</i>}
+    default-block-strategy -- "Atomic" --> merge-default-block-into-p-finish
+    default-block-strategy -- "Merge" --> default-p-r-exists
+    subgraph for-each-default-policy-rule["For each <i>r in B<i>"]
+        default-p-r-exists{"<i>p[r.id].exists?</i>"}
+        default-p-r-exists -- "Yes" --> default-p-r-exists
+        default-p-r-exists -- "No" --> default-replace-pr["<i>p[r.id] ← r</i>"] --> default-p-r-exists
+    end
+    for-each-default-policy-rule -->
+    merge-default-block-into-p-finish
+```
+
+### Merge an `overrides` block of policy rules into a policy
+
+```mermaid
+%%{ init: { "theme": "neutral" } }%%
+flowchart LR
+    merge-override-block-into-p-start([Merge override block<br>of policy rules <i>B</i><br>into policy <i>p</i>]) -->
+    override-block-strategy{<i>B.strategy</i>}
+    override-block-strategy -- "Atomic" --> full-replace-p-with-override-block[<i>p.rules ← B</i>] --> merge-override-block-into-p-finish(((Return <i>p</i>)))
+    override-block-strategy -- "Merge" --> override-replace-pr
+    subgraph for-each-override-policy-rule["For each <i>r in B<i>"]
+        override-replace-pr["<i>p[r.id] ← r</i>"] --> override-replace-pr
+    end
+    for-each-override-policy-rule -->
+    merge-override-block-into-p-finish
+```
+
+## Implementation tiers
+
+This section proposes a possible path for the implementation of this RFC for Kuadrant's existing kinds of policies that are affected by D/O – notably AuthPolicy and RateLimitPolicy.
+
+The path is divided in 3 tiers that could be delivered in steps, additionaly to a series of enhancements & refactoring.
+
+### Tier 1
+
+- Atomic defaults (currently supported; missing addition of the `defaults` field to the APIs)
+- Atomic overrides
+- Policy status and Policy discoverability (i.e. PolicyAffected status on target objects)
+- CRD labels `gateway.networking.k8s.io/policy: inherited | direct`
+
+### Tier 2
+
+- D/O `when` conditions (and support for "constraints")
+- Merge strategy
```

Comment:
I don't where to best put this comment other than here, but are the two "features" usable in combination? 
Would the strategy be conditional on "when"? Also, can I have multiple "when" clauses (if only thru multiple policies), mapping to a "switch case" of some kind? Or would we prohibit this somehow?

Reply:
The entire block (of defaults or overrides) would be conditional on `when`. I.e. if the condition evaluates to true, then apply the rules specified in the block according to the strategy specified (`atomic` or `merge`.)

TBH, I hadn't thought about multiple cases of conditions (`if..else` or `switch`-like) before. But I can see how the current proposal can be limiting for constraints. For example, how to set a constraint for a rate limit between 50 and 100? You'd need 2 blocks of overrides, though only 1 is allowed per policy; therefore you'd need 2 policies targeting the same object. Any other idea?